src/binary_annotator.cpp

```diff
@@ -0,0 +1,768 @@
+  // The returned offset will point to the root_table location.
+  const uint64_t root_table_offset = BuildHeader(0);
+
+  // Build the root table, and all else will be referenced from it.
+  BuildTable(root_table_offset, BinarySectionType::RootTable,
+             schema_->root_table());
+
+  // Now that all the sections are built, scan the regions between them and
+  // insert padding bytes that are implied.
+  FixMissingSections();
+
+  return sections_;
+}
+
+uint64_t BinaryAnnotator::BuildHeader(const uint64_t offset) {
+  std::vector<BinaryRegion> regions;
+
+  // TODO(dbaileychess): sized prefixed value
+  const uint32_t root_table_offset = GetScalar<uint32_t>(offset);
+  regions.emplace_back(MakeBinaryRegion(
+      offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0, root_table_offset,
+      std::string("offset to root table `") +
+          schema_->root_table()->name()->str() + "`"));
+
+  if (IsNonZeroRegion(offset, 4, binary_)) {
+    // Check if the file identifier region has non-zero data, and assume its the
+    // file identifier. Otherwise, it will get filled in with padding later.
+    regions.emplace_back(MakeBinaryRegion(
+        offset + sizeof(uint32_t), 4 * sizeof(uint8_t), BinaryRegionType::Char,
+        4, 0, std::string("File Identifier")));
+  }
+
+  sections_.insert(std::make_pair(
+      offset, MakeBinarySection("", BinarySectionType::Header, regions)));
+  return root_table_offset;
+}
+
+void BinaryAnnotator::BuildVTable(uint64_t offset,
+                                  const reflection::Object *table) {
+  const uint64_t vtable_offset = offset;
+
+  // First see if we have used this vtable before, if so skip building it again.
+  auto it = vtables_.find(vtable_offset);
+  if (it != vtables_.end()) { return; }
+
+  std::vector<BinaryRegion> regions;
+
+  // Vtables start with the size of the vtable
+  const uint16_t vtable_size = GetScalar<uint16_t>(offset);
+  regions.emplace_back(MakeBinaryRegion(offset, sizeof(uint16_t),
+                                        BinaryRegionType::Uint16, 0, 0,
+                                        std::string("size of this vtable")));
+  offset += sizeof(uint16_t);
+
+  // Then they have the size of the table they reference.
+  const uint16_t table_size = GetScalar<uint16_t>(offset);
+  regions.emplace_back(
+      MakeBinaryRegion(offset, sizeof(uint16_t), BinaryRegionType::Uint16, 0, 0,
+                       std::string("size of referring table")));
+  offset += sizeof(uint16_t);
+
+  const uint64_t offset_start = offset;
+
+  // A mapping between field (and its id) to the relative offset (uin16_t) from
+  // the start of the table.
+  std::map<uint16_t, VTable::Entry> fields;
+
+  // Counter for determining if the binary has more vtable entries than the
+  // schema provided. This can occur if the binary was created at a newer schema
+  // version and is being processed with an older one.
+  uint16_t fields_processed = 0;
```

Comment:
optional nit: I think it might be cleaner to combine the ForAllFields loop and the later loop for handling unknown fields into one loop from 0 to `table_size`

Reply:
This isn't easy to do due to how the ForAllFields loop works (by ID, not offset).